src/nvim/ex_docmd.c

```diff
@@ -9256,6 +9269,19 @@ put_view (
           || put_eol(fd) == FAIL) {
         return FAIL;
       }
+
+			if (fputs("call add(buffer_names, [bufname('%'),'", fd) < 0
+					|| ses_fname(fd, wp->w_buffer, flagp, false) == FAIL
```

Comment:
Must not use `ses_fname` here, it does the equivalent of `fnameescape()` which is not correct for double quotes. You need something like `string()` which will also escape NL, or like `json_encode()` which will work with non-unicode strings; I know both are or may be made (second is static inline `convert_to_json_string()`) accessible from C, but do not know whether there actually is anything which may safely dump string to file.

Reply:
Just tested, it appears that `ses_fname` is buggy also for intented purpose: can’t work with files with newlines.

Reply:
Since `ses_fname` is buggy and use everywhere in code , should it be fixed for the two case or to change it everywhere?

Reply:
@AlexandreDubray Even though it is buggy for intended purpose, intended purpose still is not valid for expressions. As it is named `ses_fname` you can overload it for expressions going into file (without renaming, I mean), but I do not think that string escaping suitable for dumping into sourced file would always be useful for sessions only, better have separate function. Or overload `convert_to_json_string`, code is going to be very similar in any case. You would need to uninline it, make not static, add boolean argument (it must not error out, dumping any possible string instead: file names with non-unicode bytes are valid, but not JSON strings) and rename to something like `encode_convert_to_double_quoted_string`.

Reply:
Ah, and it also should not use surrogate pairs to dump characters which have too big unicode codepoint number, unlike JSON VimL has `\U01234567` for that.

Reply:
@ZyX-I Oh ok, I think I misunderstood the initial problem. So what would be the best (safest) way to write the buffer name into the session file?